cmd/tpm2_autounlock/tpm2_autounlock.sh

```diff
@@ -0,0 +1,176 @@
+	shift
+done
+
+# Use -r or --rootfs value if specified, otherwise try to detect running rootfs
+if [ -z "$rootfs" ]; then
+	rootfs=$(zfs mount |awk '$2 ==  "/" { print $1 }')
+	if [ -z "$rootfs" ]; then
+		echo "Rootfs not specified (-r / --rootfs) and current rootfs could not be determined"
+		exit 1
+	fi
+fi
+
+# With rootfs determined, find the encryptionroot and its guid
+ENCRYPTIONROOT=$(zfs get -H -o value encryptionroot "${rootfs}")
+rootguid=$(zfs get guid -o value -H "${ENCRYPTIONROOT}")
+
+# Use -i or --index value if specified, otherwise look for existing property, and finally fallback to script default
+if [ -z "$tpmindex" ]; then
+	tpmindex=$(zfs get -H -o value org.zfsonlinux.tpm2:index "${ENCRYPTIONROOT}")
+	if [ "$tpmindex" = "-" ]; then
+		tpmindex="$defaulttpmindex"
+	fi
+fi
+
+# Use -P or --pcrs value if specified, otherwise look for existing property, and finally fallback to script default
+if [ -z "$tpmpcrs" ]; then
+	tpmpcrs=$(zfs get -H -o value org.zfsonlinux.tpm2:pcrs "${ENCRYPTIONROOT}")
+	if [ "$tpmpcrs" = "-" ]; then
+		tpmpcrs="$defaulttpmpcrs"
+	fi
+fi
+
+# If password is needed, use argument, piped, "script default" or stdin, in that order
+if [ -z "$password" ] && (( verifyonly+clearonly+nvlockonly == 0)); then
+	[[ -p /dev/stdin ]] && { mapfile -t; set -- "${MAPFILE[@]}"; set -- $@; }
+	pipepassword="$@"
+	if [ -n "$pipepassword" ]; then
+		password="$pipepassword"
+	elif [ -n "$defaultpassword" ]; then
+		password="$defaultpassword"
+	else
+		while true; do
+			read -s -p "Drive unlock password: " password
+			echo
+   			read -s -p "Confirm Password: " password2
+			echo
+  			[ "$password" = "$password2" ] && break
+    			echo "Please try again"
+		done
+	fi	
+fi
+
+# Check for invalid arguments
+if (( verifyonly+clearonly+nvlockonly > 1)) ; then
+	echo " You can only select only one of the following: verifyonly, clearonly, or nvlockonly"
+	exit 1
+fi
+
+########################################
+# Start execution
+if [ $clearonly ]; then
+	tpm2_nvundefine "$tpmindex" >/dev/null 2>&1
+	zfs inherit org.zfsonlinux.tpm2:index "${ENCRYPTIONROOT}"
+	zfs inherit org.zfsonlinux.tpm2:pcrs "${ENCRYPTIONROOT}"
+
+elif [ $verifyonly ]; then
+	echo "Attempting to read TPM index '$tpmindex' locked with pcrs '$tpmpcrs' and guid '$rootguid'."
+	define_policy="--policy"
+	cmd="tpm2_nvread ${tpmindex} --auth=session:s.dat+${rootguid}"
+	tpm_session_execute
+	if [ $? = 0 ]; then
+		echo "is the stored password."
+#	else
+		# The tpm2_nvread stdout messages will be visible to user
+	fi
+
+elif [ $nvlockonly ]; then
+	define_policy="--policy"
+	cmd="tpm2_nvreadlock  ${tpmindex} --auth=session:s.dat+${rootguid}"
+	tpm_session_execute
+else
+	# Preemptively clear the tpm index
+	tpm2_nvundefine "$tpmindex" >/dev/null 2>&1
+
+	# Define NVRAM location and its access rules
+	cmd="tpm2_nvdefine ${tpmindex} -Q --hierarchy=o --index-auth=${rootguid} --size=512 --policy=policy.dat --attributes='policyread|policywrite|read_stclear'"
```

Comment:
Why are you setting --index-auth without attribute TPMA_NV_AUTHREAD? IIUC, that essentially is saying, hey this is another password but i never want to use it. I would imagine the rootguid is not a secret, and thus you really just want to drop this option.

Reply:
Yes, seems redundant.  "policyread and policywrite" are the only allowed access methods. The password requirement is built within the policy and is not freestanding.

Reply:
Well..local testing didn't like removing `--index-auth=${rootguid}`.  It resulted in:
~~~
WARNING:esys:src/tss2-esys/api/Esys_NV_Write.c:306:Esys_NV_Write_Finish() Received TPM Error 
ERROR:esys:src/tss2-esys/api/Esys_NV_Write.c:110:Esys_NV_Write() Esys Finish ErrorCode (0x0000098e) 
ERROR: Failed to write NV area at index 0x1500016
ERROR: Tss2_Sys_NV_Write(0x98E) - tpm:session(1):the authorization HMAC check failed and DA counter incremented
ERROR: Unable to run tpm2_nvwrite
~~~
When running `tpm2_autounlock -p password`.  Looks like it stays.

Reply:
Probably because you still have policypassword present, which doesn't make a whole lot sense with a known value. For something like an FS encryption key, you would probably want to tie it to:
1. PCR state
2. User supplied secret (optional yet encouraged, depends on what your doing)
3. Recovery from owner hierarchy auth if user forgets passphrase or PCR state changes

Typically folks avoid hard-coding PCR's due to brittleness and use policyauthorize which says, use any policy signed by key X. Not sure what prevents rollback to a bad policy or OS state on that method, may be worth investigating. There might not be a mechanism there.. so maybe owner auth and re-creating the nv index will work for upgrade path.

The question is, for automount, where can we get the user password. Not sure where you can source it, from PAM, login session, kernel keyring, etc. I really don't know much about that side of the house and don't want to give you bad advice there. One would want a secret that is tethered to the lifetime of a successful login or from the user typing it in directly.

Item 1 - Tie to a PCR value only:
```bash
tpm2_startauthsession -S session.ctx
tpm2_policypcr -S session.ctx --pcr-list="sha256:0,1,2,3,8,9" --policy=policy.dat
tpm2_flushcontext session.ctx

tpm2_nvdefine -C o --size=512 --policy=policy.dat --attributes='policyread|policywrite|read_stclear' $NV_INDEX

tpm2_startauthsession --policy -S session.ctx
tpm2_policypcr -S session.ctx --pcr-list="sha256:0,1,2,3,8,9" 

echo 'secret' | tpm2_nvwrite -P'session:session.ctx' -i- $NV_INDEX
tpm2_flushcontext session.ctx

# should work
tpm2_startauthsession --policy -S session.ctx
tpm2_policypcr -S session.ctx --pcr-list="sha256:0,1,2,3,8,9"
tpm2_nvread -P'session:session.ctx' -s6 $NV_INDEX
tpm2_flushcontext session.ctx
```

Item 2 - tie to PCR and user password:
```bash
NV_INDEX=0x01800000

set -e -o pipefail

tpm2_startauthsession -S session.ctx
tpm2_policypcr -S session.ctx --pcr-list="sha256:0,1,2,3,8,9" 
tpm2_policypassword -S session.ctx --policy=policy.dat
tpm2_flushcontext session.ctx

tpm2_nvdefine -C o -p 'userpass' --size=512 --policy=policy.dat --attributes='policyread|policywrite|read_stclear' $NV_INDEX

tpm2_startauthsession --policy -S session.ctx
tpm2_policypcr -S session.ctx --pcr-list="sha256:0,1,2,3,8,9" 
tpm2_policypassword -S session.ctx 

echo 'secret' | tpm2_nvwrite -P'session:session.ctx+userpass' -i- $NV_INDEX
tpm2_flushcontext session.ctx

tpm2_startauthsession --policy -S session.ctx
tpm2_policypcr -S session.ctx --pcr-list="sha256:0,1,2,3,8,9"
tpm2_policypassword -S session.ctx
tpm2_nvread -P'session:session.ctx+userpass' -s6 $NV_INDEX
tpm2_flushcontext session.ctx
```
Item 3, add in recovery option.
```bash
#!/usr/bin/env bash

NV_INDEX=0x01800000

set -e -o pipefail

# this may not work for you, my TPM currently has an empty owner password
tpm2_changeauth -c o "ownerpass"

tpm2_startauthsession -S session.ctx
tpm2_policypcr -S session.ctx --pcr-list="sha256:0,1,2,3,8,9"
tpm2_policypassword -S session.ctx --policy=policy.dat
tpm2_flushcontext session.ctx

tpm2_nvdefine -C o -P 'ownerpass' -p 'userpass' --size=512 --policy=policy.dat --attributes='ownerread|policyread|policywrite|read_stclear' $NV_INDEX

tpm2_startauthsession --policy -S session.ctx
tpm2_policypcr -S session.ctx --pcr-list="sha256:0,1,2,3,8,9"
tpm2_policypassword -S session.ctx

echo 'secret' | tpm2_nvwrite -P'session:session.ctx+userpass' -i- $NV_INDEX
tpm2_flushcontext session.ctx

# recovery with owner password
tpm2_nvread -C o -P'ownerpass' -s6 $NV_INDEX

# normal flow, user auth
tpm2_startauthsession --policy -S session.ctx
tpm2_policypcr -S session.ctx --pcr-list="sha256:0,1,2,3,8,9"
tpm2_policypassword -S session.ctx
tpm2_nvread -P'session:session.ctx+userpass' -s6 $NV_INDEX
tpm2_flushcontext session.ctx
```

Reply:
Also, in my testing I found that when when you couple auth to a permanent object like owner hiearchy DA protections stop applying unless you authorize to the NV index and specify a bad authvalue.

ie
- `tpm2_nvread -Co -P garbage $NV_INDEX` will not trigger DA protections.
- `tpm2_nvread -P badsession.ctx+badpass $NV_INDEX` will not trigger DA protections.
- `tpm2_nvread -P session.ctx+badpass $NV_INDEX` will trigger DA implications and protections.

I started looking at using policysecret to couple to the owner hiearchy via a policy OR to see if that would work, but I was unable to get the script to run and ran out of time, sharing here so you can perhaps figure it out. The manpage and test code for tpm2_policyor sucks, so I will bug @idesai to look at this.

```bash
#!/usr/bin/env bash

NV_INDEX=0x01800000

set -e -o pipefail

# this may not work for you, my TPM currently has an empty owner password
tpm2_changeauth -c o "ownerpass"

do_session() {
  # PCR + user auth
  tpm2_startauthsession -S session.ctx
  tpm2_policypcr -S session.ctx --pcr-list="sha256:0,1,2,3,8,9"
  tpm2_policypassword -S session.ctx --policy=policy1.dat
  tpm2_flushcontext session.ctx

  # Tether to owner hiearchy auth, but not via attr so DA works
  tpm2_startauthsession -S session.ctx
  tpm2_policysecret -S session.ctx -c o --policy=policy2.dat 'ownerpass'
  tpm2_flushcontext session.ctx

  # create the OR policy of them both
  tpm2_startauthsession -S session.ctx
  tpm2_policyor -S session.ctx -lpolicy1.dat,policy2.dat -L policy.dat
}

do_session_user() {
  # PCR + user auth
  tpm2_startauthsession -S session.ctx --policy
  tpm2_policypcr -S session.ctx --pcr-list="sha256:0,1,2,3,8,9"
  tpm2_policypassword -S session.ctx --policy=policy1.dat
  tpm2_flushcontext session.ctx

  # create the OR policy of them both
  tpm2_startauthsession -S session.ctx --policy
  tpm2_policyor -S session.ctx -lpolicy1.dat,policy2.dat
}

do_session
tpm2_nvdefine -C o -P 'ownerpass' -p 'userpass' --size=512 --policy=policy.dat --attributes='policyread|policywrite|read_stclear' $NV_INDEX
tpm2_flushcontext session.ctx

do_session_user
echo 'secret' | tpm2_nvwrite -P'session.ctx:session.ctx+userpass' -i- $NV_INDEX
tpm2_flushcontext session.ctx

# recovery with owner password
do_session_owner
tpm2_nvread -C o -P'ownerpass' -s6 $NV_INDEX
```

Reply:
> Also, in my testing I found that when when you couple auth to a permanent object like owner hiearchy DA protections stop applying unless you authorize to the NV index and specify a bad authvalue.
> 
> ie
> 
> * `tpm2_nvread -Co -P garbage $NV_INDEX` will not trigger DA protections.
> * `tpm2_nvread -P badsession.ctx+badpass $NV_INDEX` will not trigger DA protections.
> * `tpm2_nvread -P session.ctx+badpass $NV_INDEX` will trigger DA implications and protections.
> 
> I started looking at using policysecret to couple to the owner hiearchy via a policy OR to see if that would work, but I was unable to get the script to run and ran out of time, sharing here so you can perhaps figure it out. The manpage and test code for tpm2_policyor sucks, so I will bug @idesai to look at this.
> 
> ```shell
> #!/usr/bin/env bash
> 
> NV_INDEX=0x01800000
> 
> set -e -o pipefail
> 
> # this may not work for you, my TPM currently has an empty owner password
> tpm2_changeauth -c o "ownerpass"
> 
> do_session() {
>   # PCR + user auth
>   tpm2_startauthsession -S session.ctx
>   tpm2_policypcr -S session.ctx --pcr-list="sha256:0,1,2,3,8,9"
>   tpm2_policypassword -S session.ctx --policy=policy1.dat
>   tpm2_flushcontext session.ctx
> 
>   # Tether to owner hiearchy auth, but not via attr so DA works
>   tpm2_startauthsession -S session.ctx
>   tpm2_policysecret -S session.ctx -c o --policy=policy2.dat 'ownerpass'
>   tpm2_flushcontext session.ctx
> 
>   # create the OR policy of them both
>   tpm2_startauthsession -S session.ctx
>   tpm2_policyor -S session.ctx -lpolicy1.dat,policy2.dat -L policy.dat
> }
> 
> do_session_user() {
>   # PCR + user auth
>   tpm2_startauthsession -S session.ctx --policy
>   tpm2_policypcr -S session.ctx --pcr-list="sha256:0,1,2,3,8,9"
>   tpm2_policypassword -S session.ctx --policy=policy1.dat
>   tpm2_flushcontext session.ctx
> 
>   # create the OR policy of them both
>   tpm2_startauthsession -S session.ctx --policy
>   tpm2_policyor -S session.ctx -lpolicy1.dat,policy2.dat
> }
> 
> do_session
> tpm2_nvdefine -C o -P 'ownerpass' -p 'userpass' --size=512 --policy=policy.dat --attributes='policyread|policywrite|read_stclear' $NV_INDEX
> tpm2_flushcontext session.ctx
> 
> do_session_user
> echo 'secret' | tpm2_nvwrite -P'session.ctx:session.ctx+userpass' -i- $NV_INDEX
> tpm2_flushcontext session.ctx
> 
> # recovery with owner password
> do_session_owner
> tpm2_nvread -C o -P'ownerpass' -s6 $NV_INDEX
> ```

IIUC the requirement is that, 
1. We need an authentication policy for a sealing object that if fails should trigger the DA mechanism
2. The authentication policy should be one of two — 1. PCR state 2. Recovery Password
3. PCR state should have rollback protection
4. TPM2_RH_OWNER auth should be used to update new PCR states and or the recovery password.

Reply:
@williamcroberts @ghfields this is one way to do what you are trying to achieve with enhanced authorization
 
# NV Index always read policy
tpm2_startauthsession -S session.ctx
tpm2_policycountertimer -S session.ctx -L some.policy safe
tpm2_flushcontext session.ctx
tpm2_nvdefine -C o -s 34 -a "ownerwrite|policyread|authread" -L some.policy \
-p recoverypass

# PolicySecret NV Index --> Recovery Password == NV Index Password
dd if=/dev/urandom bs=1 count=34 status=none | tpm2_nvwrite -C o -i- 0x01000000
tpm2_startauthsession -S session.ctx
tpm2_policysecret -S session.ctx -L policy.secret.nvindex \
-c 0x01000000 recoverypass
tpm2_flushcontext session.ctx 

# PolicyPCR
tpm2_startauthsession -S session.ctx
tpm2_policypcr -S session.ctx -l "sha1:0,1,2,3" -L policy.pcr
tpm2_flushcontext session.ctx

# Policy ( PCR || Recovery-Password)
tpm2_startauthsession -S session.ctx 
tpm2_policyor -S session.ctx -l sha256:policy.pcr,policy.secret.nvindex \
-L policy.or
tpm2_flushcontext session.ctx

# Policy written out to NV index with owner auth.
# New PCR states will replace this value.
echo "000b" | xxd -p -r | cat - policy.or | tpm2_nvwrite -C o -i- 0x01000000
tpm2_startauthsession -S session.ctx
tpm2_policyauthorizenv -S session.ctx -L policy.authnv -C 0x01000000 0x01000000 \
-P recoverypass
tpm2_flushcontext session.ctx

# Sealing object
tpm2_createprimary -C o -c prim.ctx -Q
echo "idesaisecret"| tpm2_create -Q -C prim.ctx -u key.pub -r key.priv \
-a "fixedtpm|fixedparent|adminwithpolicy" -L policy.authnv -i-
tpm2_load -C prim.ctx -u key.pub -r key.priv -c key.ctx

# Unsealing with PolicyPCR - Failing this triggers DA mechanism
tpm2_startauthsession -S session_read.ctx --policy-session
tpm2_policycountertimer -S session_read.ctx safe
tpm2_startauthsession -S session.ctx --policy-session
tpm2_policypcr -S session.ctx -l "sha1:0,1,2,3"
tpm2_policyor -S session.ctx -l sha256:policy.pcr,policy.secret.nvindex
tpm2_policyauthorizenv -S session.ctx -C 0x01000000 0x01000000 \
-P session:session_read.ctx
tpm2_unseal  -c  key.ctx  -p session:session.ctx | xxd -p
tpm2_flushcontext session.ctx

# Unsealing with Recovery-Password - Failing this triggers DA mechanism
tpm2_startauthsession -S session_read.ctx --policy-session
tpm2_policycountertimer -S session_read.ctx safe
tpm2_startauthsession -S session.ctx --policy-session
tpm2_policysecret -S session.ctx -L policy.secret.nvindex \
-c 0x01000000 recoverypass
tpm2_policyor -S session.ctx -l sha256:policy.pcr,policy.secret.nvindex
tpm2_policyauthorizenv -S session.ctx -C 0x01000000 0x01000000 \
-P session:session_read.ctx
tpm2_unseal  -c  key.ctx  -p session:session.ctx | xxd -p
tpm2_flushcontext session.ctx